File: scripts/geocode_new_regions.mjs
#!/usr/bin/env node
/**
 * geocode_new_regions.mjs
 *
 * Targeted Nominatim geocoding for 5 newly created regions:
 * - Niagara Peninsula (Canada)
 * - Okanagan Valley (Canada)
 * - Klein Karoo (South Africa)
 * - Olifants River (South Africa)
 * - Scotland (UK)
 *
 * Usage:
 *   node scripts/geocode_new_regions.mjs --dry-run
 *   node scripts/geocode_new_regions.mjs --apply
 */

import { readFileSync } from 'fs';
import { createClient } from '@supabase/supabase-js';

// ── Load .env ───────────────────────────────────────────────
const envPath = new URL('../.env', import.meta.url).pathname.replace(/^\/([A-Z]:)/, '$1');
const envContent = readFileSync(envPath, 'utf-8');
for (const line of envContent.split('\n')) {
  const trimmed = line.replace(/\r/g, '').trim();
  if (!trimmed || trimmed.startsWith('#')) continue;
  const eqIdx = trimmed.indexOf('=');
  if (eqIdx === -1) continue;
  const key = trimmed.slice(0, eqIdx);
  const val = trimmed.slice(eqIdx + 1);
  if (!process.env[key]) process.env[key] = val;
}

const SUPABASE_URL = process.env.SUPABASE_URL;
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_ROLE || process.env.SUPABASE_ANON_KEY;
const sb = createClient(SUPABASE_URL, SUPABASE_KEY);

const APPLY = process.argv.includes('--apply');
const MODE = APPLY ? 'APPLY' : 'DRY RUN';

const NOMINATIM_URL = 'https://nominatim.openstreetmap.org/search';
const USER_AGENT = 'Loam/1.0 (wine-intelligence-platform)';

function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

async function nominatimSearch(query) {
  const params = new URLSearchParams({
    q: query, format: 'jsonv2', limit: '1', polygon_geojson: '1',
  });
  const res = await fetch(`${NOMINATIM_URL}?${params}`, {
    headers: { 'User-Agent': USER_AGENT },
  });
  if (!res.ok) throw new Error(`Nominatim ${res.status}`);
  const results = await res.json();
  return results[0] || null;
}

function simplifyPrecision(geojson) {
  function roundCoords(coords) {
    if (typeof coords[0] === 'number') {
      return [Math.round(coords[0] * 100000) / 100000, Math.round(coords[1] * 100000) / 100000];
    }
    return coords.map(c => roundCoords(c));
  }
  return { type: geojson.type, coordinates: roundCoords(geojson.coordinates) };
}

// Targeted queries for each region
const REGIONS = [
  {
    slug: 'niagara-peninsula-region',
    queries: ['Niagara Peninsula, Ontario, Canada', 'Niagara Region, Ontario, Canada'],
  },
  {
    slug: 'okanagan-valley-region',
    queries: ['Okanagan Valley, British Columbia, Canada', 'Regional District of Central Okanagan, British Columbia'],
  },
  {
    slug: 'klein-karoo-region',
    queries: ['Klein Karoo, South Africa', 'Little Karoo, South Africa', 'Klein Karoo, Western Cape'],
  },
  {
    slug: 'olifants-river-region',
    queries: ['Olifants River Valley, South Africa', 'Citrusdal, Western Cape, South Africa'],
  },
  {
    slug: 'scotland',
    queries: ['Scotland, United Kingdom', 'Scotland'],
  },
];

async function main() {
  console.log(`\n=== Geocode New Regions (${MODE}) ===\n`);

  // Get region IDs
  const slugs = REGIONS.map(r => r.slug);
  const { data: regions, error: regErr } = await sb
    .from('regions')
    .select('id, slug, name')
    .in('slug', slugs)
    .is('deleted_at', null);
  if (regErr) throw regErr;

  const regionBySlug = {};
  for (const r of regions) regionBySlug[r.slug] = r;

  let fixed = 0;
  let failed = 0;

  for (const { slug, queries } of REGIONS) {
    const region = regionBySlug[slug];
    if (!region) {
      console.log(`  ⚠️ Region not found: ${slug}`);
      continue;
    }

    console.log(`  ${region.name} (${slug})`);

    if (!APPLY) {
      console.log(`    Would try: ${queries.join(' → ')}`);
      continue;
    }

    let result = null;
    for (const query of queries) {
      await sleep(1100);
      console.log(`    Trying: "${query}"`);
      result = await nominatimSearch(query);
      if (result) {
        const isPolygon = result.geojson &&
          (result.geojson.type === 'Polygon' || result.geojson.type === 'MultiPolygon');
        if (isPolygon) break; // Found a polygon, use it
        console.log(`    Got centroid only, trying next query...`);
        // Keep this result as fallback but try for a polygon
      }
    }

    if (!result) {
      console.log(`    ❌ All queries failed`);
      failed++;
      continue;
    }

    const lat = parseFloat(result.lat);
    const lng = parseFloat(result.lon);
    const geojson = result.geojson;
    const osmType = result.osm_type;
    const osmId = result.osm_id;
    const sourceId = `nominatim/${osmType}/${osmId}`;
    const isPolygon = geojson && (geojson.type === 'Polygon' || geojson.type === 'MultiPolygon');

    if (isPolygon) {
      const simplified = simplifyPrecision(geojson);
      const sizeKB = (JSON.stringify(simplified).length / 1024).toFixed(1);
      console.log(`    ✅ Polygon (${simplified.type}, ${sizeKB} KB)`);

      const { error: rpcErr } = await sb.rpc('upsert_region_boundary', {
        p_region_id: region.id,
        p_geojson: JSON.stringify(simplified),
        p_source_id: sourceId,
        p_confidence: 'approximate',
      });
      if (rpcErr) { console.log(`    ❌ RPC: ${rpcErr.message}`); failed++; continue; }
    } else {
      console.log(`    📍 Centroid (${lat.toFixed(4)}, ${lng.toFixed(4)})`);

      const { error: rpcErr } = await sb.rpc('upsert_region_boundary', {
        p_region_id: region.id,
        p_lat: lat,
        p_lng: lng,
        p_source_id: sourceId,
        p_confidence: 'geocoded',
      });
      if (rpcErr) { console.log(`    ❌ RPC: ${rpcErr.message}`); failed++; continue; }
    }
    fixed++;
  }

  console.log(`\n=== Summary ===`);
  console.log(`Fixed: ${fixed}`);
  console.log(`Failed: ${failed}`);
  console.log(`\nDone!`);
}

main().catch(e => { console.error(e); process.exit(1); });

File: pipeline/geo/geocode_new_regions.py
"""
Targeted Nominatim geocoding for specific newly created regions.

Usage:
  python -m pipeline.geo.geocode_new_regions --dry-run
  python -m pipeline.geo.geocode_new_regions --apply
"""

import sys
import json
import time
import argparse
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

import httpx
from pipeline.lib.db import get_supabase
from pipeline.geo.helpers import nominatim_search, simplify_precision, RATE_LIMIT_S

REGIONS = [
    {"slug": "niagara-peninsula-region",
     "queries": ["Niagara Peninsula, Ontario, Canada", "Niagara Region, Ontario, Canada"]},
    {"slug": "okanagan-valley-region",
     "queries": ["Okanagan Valley, British Columbia, Canada", "Regional District of Central Okanagan, British Columbia"]},
    {"slug": "klein-karoo-region",
     "queries": ["Klein Karoo, South Africa", "Little Karoo, South Africa", "Klein Karoo, Western Cape"]},
    {"slug": "olifants-river-region",
     "queries": ["Olifants River Valley, South Africa", "Citrusdal, Western Cape, South Africa"]},
    {"slug": "scotland",
     "queries": ["Scotland, United Kingdom", "Scotland"]},
]


def main():
    parser = argparse.ArgumentParser(description="Geocode new regions")
    parser.add_argument("--apply", action="store_true")
    parser.add_argument("--dry-run", action="store_true")
    args = parser.parse_args()

    apply = args.apply
    sb = get_supabase()
    client = httpx.Client(timeout=30)
    mode = "APPLY" if apply else "DRY RUN"
    print(f"\n=== Geocode New Regions ({mode}) ===\n")

    slugs = [r["slug"] for r in REGIONS]
    result = sb.table("regions").select("id, slug, name").in_("slug", slugs).is_(
        "deleted_at", "null"
    ).execute()
    region_by_slug = {r["slug"]: r for r in result.data}

    fixed = 0
    failed = 0

    for entry in REGIONS:
        region = region_by_slug.get(entry["slug"])
        if not region:
            print(f"  WARNING Region not found: {entry['slug']}")
            continue

        print(f"  {region['name']} ({entry['slug']})")

        if not apply:
            print(f"    Would try: {' -> '.join(entry['queries'])}")
            continue

        result_data = None
        for query in entry["queries"]:
            time.sleep(RATE_LIMIT_S)
            print(f"    Trying: \"{query}\"")
            try:
                results = nominatim_search(client, query)
                if results:
                    r = results[0]
                    geojson = r.get("geojson")
                    is_poly = geojson and geojson.get("type") in ("Polygon", "MultiPolygon")
                    if is_poly:
                        result_data = r
                        break
                    result_data = r  # keep as fallback
            except Exception as e:
                print(f"    Error: {e}")

        if not result_data:
            print(f"    FAIL All queries failed")
            failed += 1
            continue

        lat = float(result_data["lat"])
        lng = float(result_data["lon"])
        geojson = result_data.get("geojson")
        source_id = f"nominatim/{result_data.get('osm_type')}/{result_data.get('osm_id')}"
        is_poly = geojson and geojson.get("type") in ("Polygon", "MultiPolygon")

        if is_poly:
            simplified = simplify_precision(geojson)
            size_kb = len(json.dumps(simplified)) / 1024
            print(f"    ok Polygon ({simplified['type']}, {size_kb:.1f} KB)")
            sb.rpc("upsert_region_boundary", {
                "p_region_id": region["id"],
                "p_geojson": json.dumps(simplified),
                "p_source_id": source_id,
                "p_confidence": "approximate",
            }).execute()
        else:
            print(f"    ok Centroid ({lat:.4f}, {lng:.4f})")
            sb.rpc("upsert_region_boundary", {
                "p_region_id": region["id"],
                "p_lat": lat, "p_lng": lng,
                "p_source_id": source_id,
                "p_confidence": "geocoded",
            }).execute()
        fixed += 1

    client.close()
    print(f"\n=== Summary: Fixed={fixed}, Failed={failed} ===")


if __name__ == "__main__":
    main()
